memcache/selector.go

```diff
@@ -111,19 +153,142 @@ var keyBufPool = sync.Pool{
 	},
 }
 
-func (ss *ServerList) PickServer(key string) (net.Addr, error) {
-	ss.mu.RLock()
-	defer ss.mu.RUnlock()
-	if len(ss.addrs) == 0 {
+func (ss *ServerList) PickServer(key string) (picked net.Addr, err error) {
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
+
+	// if we pick a connection for retry, remove it from the available pool till we
+	// hear the result
+	defer (func() {
+		if err != nil {
+			return
+		}
+		// remove the address from available roster again to avoid multiple retries
+		if ws, ok := ss.states[picked]; ok && ws.retry == retryReady {
+			ws.retry = retryRunning
+			ss.setState(picked, ws)
+		}
+	})()
+
+	if len(ss.available) == 0 {
 		return nil, ErrNoServers
 	}
-	if len(ss.addrs) == 1 {
-		return ss.addrs[0], nil
+	if len(ss.available) == 1 {
+		return ss.available[0], nil
 	}
 	bufp := keyBufPool.Get().(*[]byte)
 	n := copy(*bufp, key)
 	cs := crc32.ChecksumIEEE((*bufp)[:n])
 	keyBufPool.Put(bufp)
 
-	return ss.addrs[cs%uint32(len(ss.addrs))], nil
+	return ss.available[cs%uint32(len(ss.available))], nil
```

Comment:
Great spot. I wonder if we should add in a consistent hashing algorithm? The base implementation also rehashed keys when `SetServers` was called, but without many comments in the docs it's unclear how that was used.

Shall we ship as is to avoid more node crashes? At least this rehashing is at least going to be less bad than the status quo, which is an unrecoverable retry storm + permanent loss of cache (as we've resolved the DNS immutably)?

Reply:
Okay that sounds good. Would be good to track this in JIRA as a "Medium" priority (DevEff lingo for "do this cycle")